tests/phpunit/tests/dependencies/scripts.php

```diff
@@ -3065,4 +3065,85 @@ protected function assertEqualMarkup( $expected, $actual, $message = '' ) {
 	protected function add_html5_script_theme_support() {
 		add_theme_support( 'html5', array( 'script' ) );
 	}
+
+	/**
+	 * Test that a script is moved to the footer if it is made non deferrable, was in the header and
+	 * all scripts that depend on it are in the footer.
+	 *
+	 * @ticket 58599
+	 *
+	 * @dataProvider data_provider_script_move_to_footer
+	 *
+	 * @param string $b_in_footer in_footer value for script B.
+	 * @param string $expected    Expected output.
+	 * @param bool   $expect_a_in_footer Is script A in the footer?
+	 * @param bool   $expect_b_in_footer Is script B in the footer?
+	 */
+	public function test_wp_scripts_move_to_footer( $b_in_footer, $expected, $expect_a_in_footer, $expect_b_in_footer ) {
+		// Script A uses strategy=>defer. Script B depends on script A and uses in_footer=>true.
+		wp_enqueue_script( 'script-a', 'script-a.js', array(), null, array( 'strategy' => 'defer' ) );
+		wp_enqueue_script( 'script-b', 'script-b.js', array( 'script-a' ), null, $b_in_footer );
+
+		// Run the header and footer capturing the output.
+		ob_start();
+		wp_scripts()->do_head_items();
+		wp_scripts()->do_footer_items();
+		$actual = ob_get_clean();
```

Comment:
I think these should be captured separately, so we can assert whether a script ended up in the head or the footer. Otherwise, we can't tell from the test where they were printed, except for looking at `in_footer` which I think shouldn't be relied on by itself.

Reply:
👍🏼 I will work to improve the tests with that and also I'm going to add a script c that depends on script a to verify that if it is in the header, script a is not moved

Reply:
broke these out